File: services/api/handlers_health.go
package main

import (
	"context"
	"encoding/json"
	"net/http"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
	influxdb2 "github.com/influxdata/influxdb-client-go/v2"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/redis/go-redis/v9"
	kafka "github.com/segmentio/kafka-go"
)

// handleLive always returns 200 while the process is up — liveness must
// never fail just because a downstream dependency (Postgres, Kafka, ...) is
// temporarily unavailable, or an orchestrator would kill and restart a
// perfectly healthy process for a problem restarting it can't fix.
func handleLive() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("ok"))
	}
}

type depStatus struct {
	Postgres string `json:"postgres"`
	Redis    string `json:"redis"`
	Kafka    string `json:"kafka"`
	MQTT     string `json:"mqtt"`
	InfluxDB string `json:"influxdb"`
}

func checkDependencies(cfg Config, pool *pgxpool.Pool, redisClient *redis.Client) depStatus {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	status := depStatus{Postgres: "ok", Redis: "ok", Kafka: "ok", MQTT: "ok", InfluxDB: "ok"}

	if err := pool.Ping(ctx); err != nil {
		status.Postgres = "unreachable: " + err.Error()
	}
	if err := redisClient.Ping(ctx).Err(); err != nil {
		status.Redis = "unreachable: " + err.Error()
	}

	conn, err := kafka.DialContext(ctx, "tcp", cfg.KafkaBrokers[0])
	if err != nil {
		status.Kafka = "unreachable: " + err.Error()
	} else {
		_ = conn.Close()
	}

	opts := mqtt.NewClientOptions().AddBroker(cfg.MQTTBrokerURL).SetConnectTimeout(2 * time.Second)
	client := mqtt.NewClient(opts)
	token := client.Connect()
	if !token.WaitTimeout(2*time.Second) || token.Error() != nil {
		status.MQTT = "unreachable"
	} else {
		client.Disconnect(100)
	}

	influxClient := influxdb2.NewClient(cfg.InfluxURL, cfg.InfluxToken)
	defer influxClient.Close()
	if _, err := influxClient.Ping(ctx); err != nil {
		status.InfluxDB = "unreachable: " + err.Error()
	}

	return status
}

// handleReady fails (503) if any dependency this instance actually needs
// to serve requests is unreachable — distinct from liveness, per spec.
func handleReady(cfg Config, pool *pgxpool.Pool, redisClient *redis.Client) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		status := checkDependencies(cfg, pool, redisClient)
		ready := status.Postgres == "ok" && status.Redis == "ok" && status.Kafka == "ok" && status.InfluxDB == "ok"
		// MQTT is deliberately excluded from readiness: the API doesn't
		// itself consume/publish MQTT, so its outage shouldn't take API
		// readiness down with it — it's still reported in /health though.

		w.Header().Set("Content-Type", "application/json")
		if !ready {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"ready": ready, "dependencies": status})
	}
}

func handleHealth(cfg Config, pool *pgxpool.Pool, redisClient *redis.Client) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		status := checkDependencies(cfg, pool, redisClient)
		writeJSON(w, http.StatusOK, status)
	}
}

File: services/api/handlers_health.py
"""Health/readiness endpoints, the Python port of handlers_health.go."""

from __future__ import annotations

from dataclasses import dataclass

import paho.mqtt.client as mqtt
from confluent_kafka.admin import AdminClient
from fastapi import APIRouter, Response
from influxdb_client import InfluxDBClient
from psycopg_pool import ConnectionPool
from redis import Redis

from config import Config

router = APIRouter()


@dataclass
class DepStatus:
    postgres: str = "ok"
    redis: str = "ok"
    kafka: str = "ok"
    mqtt: str = "ok"
    influxdb: str = "ok"


def _check_dependencies(cfg: Config, pool: ConnectionPool, redis_client: Redis) -> DepStatus:
    status = DepStatus()

    try:
        with pool.connection(timeout=2.0) as conn:
            conn.execute("SELECT 1")
    except Exception as exc:  # noqa: BLE001
        status.postgres = f"unreachable: {exc}"

    try:
        redis_client.ping()
    except Exception as exc:  # noqa: BLE001
        status.redis = f"unreachable: {exc}"

    try:
        broker = cfg.kafka_brokers[0]
        admin = AdminClient({"bootstrap.servers": broker, "socket.timeout.ms": 2000})
        admin.list_topics(timeout=2.0)
    except Exception as exc:  # noqa: BLE001
        status.kafka = f"unreachable: {exc}"

    try:
        host, _, port = cfg.mqtt_broker_url.replace("tcp://", "").partition(":")
        client = mqtt.Client(callback_api_version=mqtt.CallbackAPIVersion.VERSION2)
        client.connect(host, int(port) if port else 1883, keepalive=2)
        client.loop(timeout=2.0)
        if not client.is_connected():
            status.mqtt = "unreachable"
        client.disconnect()
    except Exception:  # noqa: BLE001
        status.mqtt = "unreachable"

    try:
        influx_client = InfluxDBClient(url=cfg.influx_url, token=cfg.influx_token, org=cfg.influx_org, timeout=2_000)
        try:
            if not influx_client.ping():
                status.influxdb = "unreachable"
        finally:
            influx_client.close()
    except Exception as exc:  # noqa: BLE001
        status.influxdb = f"unreachable: {exc}"

    return status


def register_health_routes(cfg: Config, pool: ConnectionPool, redis_client: Redis) -> APIRouter:
    @router.get("/live")
    def live() -> Response:
        return Response(content="ok", status_code=200)

    @router.get("/ready")
    def ready(response: Response) -> dict:
        status = _check_dependencies(cfg, pool, redis_client)
        # MQTT is deliberately excluded from readiness: the API doesn't
        # itself consume/publish MQTT, so its outage shouldn't take API
        # readiness down with it -- it's still reported in /health though.
        ready_ = status.postgres == "ok" and status.redis == "ok" and status.kafka == "ok" and status.influxdb == "ok"
        response.status_code = 200 if ready_ else 503
        return {"ready": ready_, "dependencies": status.__dict__}

    @router.get("/health")
    def health() -> dict:
        return _check_dependencies(cfg, pool, redis_client).__dict__

    return router
